src/metabase/util/malli/registry.cljc

```diff
@@ -3,28 +3,123 @@
   (:require
    #?@(:clj ([malli.experimental.time :as malli.time]
              [net.cgrand.macrovich :as macros]))
+   [clojure.walk :as walk]
    [malli.core :as mc]
    [malli.registry]
-   [malli.util :as mut])
+   [malli.util :as mut]
+   [metabase.config.core :as config])
   #?(:cljs (:require-macros [metabase.util.malli.registry])))
 
 (defonce ^:private cache (atom {}))
 
 (defn- schema-cache-key
   "Make schemas that aren't `=` to identical ones e.g.
-
     [:re #\"\\d{4}\"]
+    [:or [:re #\"\\d{4}\"] :int]
+    [:fn even?]
+    (mr/with-key [:fn (constantly true)])
 
   work correctly as cache keys instead of creating new entries every time the code is evaluated."
   [x]
-  (if (and (vector? x)
-           (= (first x) :re))
-    (into (empty x)
-          (map (fn [child]
-                 (cond-> child
-                   (instance? #?(:clj java.util.regex.Pattern :cljs js/RegExp) child) str)))
-          x)
-    x))
+  (or (some-> x meta ::key)
+      (walk/postwalk
+       (fn [x]
+         (cond
+           (instance? #?(:clj java.util.regex.Pattern
+                         :cljs js/RegExp) x) (str x)
+           ;; Attach a cache key to any naughty object in a schema, and it'll be used
+           ;; to calculate the cache-key for that part.
+           (some-> x meta ::key) (-> x meta ::key)
+           ;; TODO: This does not work for h.o.f. like (every-pred even? odd?),
+           ;; (pr-str (every-pred even? odd?)) == (pr-str (every-pred string?))
+           ;; (fn? x) (pr-str x)
+           :else x))
+       x)))
+
+;;;; `with-key` helpers
+
+(defn- stable-sym-names
+  "Returns a lazy seq of symbols that can be used as stable argument names."
+  []
+  (letfn [(int-to-stable-arg [n]
+            (loop [num n
+                   result ""]
+              (if (zero? num)
+                (symbol result)
+                (let [remainder (mod (dec num) 26)
+                      char (char (+ (int \a) remainder))]
+                  (recur (quot (dec num) 26)
+                         (str char result))))))]
```

Comment:
This is neat but is ultimately unnecessary (IIUC). You could probably just have something like `(map #(symbol (str "arg" %)) (iterate inc 1))`, or even `(into {} (map-indexed (fn [i arg] [arg (str "arg")])))` in the following function. Not as creative, but more to the point.